src/treq/auth.py

```diff
@@ -1,13 +1,223 @@
+            digest_hash_func = _sha512_utf_digest
+        else:
+            raise UnknownDigestAuthAlgorithm(algo)
+
+        ha1 = digest_hash_func(a1)
+        ha2 = digest_hash_func(a2)
+
+        cnonce = _generate_client_nonce(nonce)
+
+        if algo == b'MD5-SESS':
+            sess = ha1.encode('utf-8')
+            sess += b':'
+            sess += nonce
+            sess += b':'
+            sess += cnonce.encode('utf-8')
+            ha1 = digest_hash_func(sess)
+
+        if cached:
+            self._digest_auth_cache[(method, path)]['c'] += 1
+            nonce_count = self._digest_auth_cache[(method, path)]['c']
+        else:
+            nonce_count = 1
+
+        ncvalue = '%08x' % nonce_count
+        if qop is None:
+            rd = ha1.encode('utf-8')
+            rd += b':'
+            rd += ha2.encode('utf-8')
+            rd += b':'
+            rd += nonce
+            response_digest = digest_hash_func(rd).encode('utf-8')
+        else:
+            rd = ha1.encode('utf-8')
+            rd += b':'
+            rd += nonce
+            rd += b':'
+            rd += ncvalue.encode('utf-8')
+            rd += b':'
+            rd += cnonce.encode('utf-8')
+            rd += b':'
+            if not isinstance(qop, bytes):
+                rd += qop.encode('utf-8')
+            else:
+                rd += qop
+            rd += b':'
+            rd += ha2.encode('utf-8')
+            response_digest = digest_hash_func(rd).encode('utf-8')
+        hb = b'username="'
+        hb += self._username
+        hb += b'", realm="'
+        hb += realm
+        hb += b'", nonce="'
+        hb += nonce
+        hb += b'", uri="'
+        hb += actual_path
+        hb += b'", response="'
+        hb += response_digest
+        hb += b'"'
+        if opaque:
+            hb += b', opaque="'
+            hb += opaque
+            hb += b'"'
+        if original_algo:
+            hb += b', algorithm="'
+            hb += original_algo
+            hb += b'"'
+        if qop:
+            hb += b', qop="'
+            if not isinstance(qop, bytes):
+                hb += qop.encode('utf-8')
+            else:
+                hb += qop
+            hb += b'", nc='
+            hb += ncvalue.encode('utf-8')
+            hb += b', cnonce="'
+            hb += cnonce.encode('utf-8')
+            hb += b'"'
+        if not cached:
+            cache_params = {
+                'path': path,
+                'method': method,
+                'cached': cached,
+                'nonce': nonce,
+                'realm': realm,
+                'qop': qop,
+                'algorithm': algorithm,
+                'opaque': opaque
+            }
+            self._digest_auth_cache[(method, path)] = {
+                'p': cache_params,
+                'c': 1
+            }
+        digest_res = b'Digest '
+        digest_res += hb
+        return digest_res
+
+    def cached_metadata_for(self, method: bytes, uri: bytes) -> Optional[dict]:
```

Comment:
Does this need to be a public method?

Reply:
Hmm, probably not, I'll change that to private.

Reply:
changed to private